In [1]:
import kaggle
import zipfile
import pandas as pd

In [2]:
# downloading a dataset
!kaggle datasets download -d rashikrahmanpritom/data-science-job-posting-on-glassdoor

Dataset URL: https://www.kaggle.com/datasets/rashikrahmanpritom/data-science-job-posting-on-glassdoor
License(s): CC0-1.0




  0%|          | 0.00/1.53M [00:00<?, ?B/s]
 65%|██████▌   | 1.00M/1.53M [00:00<00:00, 2.23MB/s]
100%|██████████| 1.53M/1.53M [00:00<00:00, 2.61MB/s]
100%|██████████| 1.53M/1.53M [00:00<00:00, 2.53MB/s]


In [3]:
# extracting a zip file
zipfile_name = 'data-science-job-posting-on-glassdoor.zip'
with zipfile.ZipFile(zipfile_name, 'r') as file:
    file.extractall()

*note*: at this oint, we have two files:

- Uncleaned_DS_jobs.csv
- Cleaned_DS_Jobs.csv

let us **get rid of** Cleaned_DS_Jobs.csv now

In [4]:
import os

In [5]:
file_to_rm = 'Cleaned_DS_Jobs.csv'
os.remove(file_to_rm)
# changing a name of a file
os.rename('Uncleaned_DS_jobs.csv', 'job_posting.csv')

In [6]:
# reading csv file as pandas dataframe
job_posting = pd.read_csv('job_posting.csv')

In [7]:
# checking our dataframe
job_posting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 672 entries, 0 to 671
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   index              672 non-null    int64  
 1   Job Title          672 non-null    object 
 2   Salary Estimate    672 non-null    object 
 3   Job Description    672 non-null    object 
 4   Rating             672 non-null    float64
 5   Company Name       672 non-null    object 
 6   Location           672 non-null    object 
 7   Headquarters       672 non-null    object 
 8   Size               672 non-null    object 
 9   Founded            672 non-null    int64  
 10  Type of ownership  672 non-null    object 
 11  Industry           672 non-null    object 
 12  Sector             672 non-null    object 
 13  Revenue            672 non-null    object 
 14  Competitors        672 non-null    object 
dtypes: float64(1), int64(2), object(12)
memory usage: 78.9+ KB


so, ``job_posting`` dataframe consists of:
- 672 rows
- 15 columns

In [8]:
# how our uncleaned dataset looks like
job_posting.head()

index          Job Title               Salary Estimate  \
0      0  Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1      1     Data Scientist  $137K-$171K (Glassdoor est.)   
2      2     Data Scientist  $137K-$171K (Glassdoor est.)   
3      3     Data Scientist  $137K-$171K (Glassdoor est.)   
4      4     Data Scientist  $137K-$171K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  
0            EmblemHealth, UnitedHealth Group, Aetna  
1                                                 -1  
2                                                 -1  
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4               Commerce Signals, Cardlytics, Yodlee

right off the bat, there are several changes we can do to make our data more presentable:
1. in *Salary Estimate* column: get rid of dollar signs and *(Glassdoor est.)*
2. in *Company Name* column: left slice everything before newline character appearance
3. in *Job Description* column: get rid of newline chars, maybe replace them with ' '(?)
4. in a whole dataframe: replace *-1* and *Unknown / Non-Applicable* values to *Unknown*, since both of those values that we want to replace basically mean the same thing
5. add the following columns based on *Salary Estimate*:
    - *Min salary*
    - *Max salary*
    - *Average salary*
6. add *Same state* column, that will represent whether or not a companys location and headquartes are located in the same state

#### Min, Max, Average Salaries

In [9]:
# rm "(Glassdoor est.)" 
job_posting['Salary Estimate'] = job_posting['Salary Estimate'].map(lambda x: x.split('(')[0])

# function that returns min and max salaries
def seperate_est(est):
    min_sal, max_sal = est.split('-')

    # removing '$' nad 'K' from a string, converting it to int and multiplying it by 1000
    min_sal = int(min_sal.strip('$').replace('K', '')) * 1000
    max_sal = int(max_sal.strip('$').replace('K', '')) * 1000

    return min_sal, max_sal

# assigning 'Min Salary' and 'Max Salary' columns
job_posting[['Min Salary', 'Max Salary']] = job_posting['Salary Estimate'].apply(lambda x: pd.Series(seperate_est(x)))

# assigning a 'Average Salary' column
job_posting['Average Salary'] = (job_posting['Min Salary'] + job_posting['Max Salary']) / 2

# reasigning a 'Salary Estimate' column
job_posting['Salary Estimate'] = job_posting['Min Salary'].astype(str) + '-' + job_posting['Max Salary'].astype(str)

Now, let us have a look at columns we worked on.

In [10]:
job_posting[['Min Salary', 'Max Salary', 'Salary Estimate', 'Average Salary']].head()

Min Salary  Max Salary Salary Estimate  Average Salary
0      137000      171000   137000-171000        154000.0
1      137000      171000   137000-171000        154000.0
2      137000      171000   137000-171000        154000.0
3      137000      171000   137000-171000        154000.0
4      137000      171000   137000-171000        154000.0

### Removing '\n' in a *Company Name* column

In *Company Name* column are mixed values.

In [11]:
job_posting['Company Name'].tail()

667                TRANZACT\n3.6
668                         JKGT
669                   AccessHope
670    ChaTeck Incorporated\n5.0
671           1-800-Flowers\n2.7
Name: Company Name, dtype: object

It looks like in some cases, rows have a syntax consisting of 'company name', '\n' and 'company rating'.

Let us extract only the name of the company.

In [12]:
job_posting['Company Name'] = job_posting['Company Name'].map(lambda x: x.split('\n')[0])
job_posting['Company Name'].tail()

667                TRANZACT
668                    JKGT
669              AccessHope
670    ChaTeck Incorporated
671           1-800-Flowers
Name: Company Name, dtype: object

Now it is all clean!

### '\n' chars inside *Job Description* column 

In [13]:
job_posting['Job Description'].head()

0    Description\n\nThe Senior Data Scientist is re...
1    Secure our Nation, Ignite your Future\n\nJoin ...
2    Overview\n\n\nAnalysis Group is one of the lar...
3    JOB DESCRIPTION:\n\nDo you have a passion for ...
4    Data Scientist\nAffinity Solutions / Marketing...
Name: Job Description, dtype: object

We need to remove every newline character in a column. However, if we use for example just a `.replace()` method, as we did before, we can be left with:

- no space between words, if we choose to replace \n with ''
- a huge gap between words, if we replace \n with ' '

So, we will have to ensure that no matter how many \n are in a string, there has to be only one space in their place.

In [14]:
# importing regex library
import re

def remove_newline(val):
    # replacing enery newline char with ' '
    val = val.replace('\n', ' ')

    # replacing any sequence of spaces with a single space
    val = re.sub(r'\s+', ' ', val)

    # returning a value with removed any possible leading or trailing spaces
    return val.strip()

# reassigning a 'Job Description' column
job_posting['Job Description'] = job_posting['Job Description'].apply(remove_newline)

In [15]:
# viewing our recently modified column
job_posting['Job Description'].head()

0    Description The Senior Data Scientist is respo...
1    Secure our Nation, Ignite your Future Join the...
2    Overview Analysis Group is one of the largest ...
3    JOB DESCRIPTION: Do you have a passion for Dat...
4    Data Scientist Affinity Solutions / Marketing ...
Name: Job Description, dtype: object

Looks magnificent!

### Replacing values in str columns to 'Unknown'

In [17]:
col_names_str = ['Job Title', 'Job Description', 'Company Name', 'Location', 'Headquarters', 'Size',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors']

job_posting[col_names_str] = job_posting[col_names_str].replace('-1', 'Unknown')
job_posting[col_names_str] = job_posting[col_names_str].replace('Unknown / Non-Applicable', 'Unknown')


job_posting.tail()

index       Job Title Salary Estimate  \
667    667  Data Scientist   105000-167000   
668    668  Data Scientist   105000-167000   
669    669  Data Scientist   105000-167000   
670    670  Data Scientist   105000-167000   
671    671  Data Scientist   105000-167000   

                                       Job Description  Rating  \
667  Summary We’re looking for a data scientist to ...     3.6   
668  Job Description Become a thought leader within...    -1.0   
669  Join a thriving company that is changing the w...    -1.0   
670  100 Remote Opportunity As an AINLP Data Scient...     5.0   
671  Description The Data Scientist will be part of...     2.7   

             Company Name           Location     Headquarters  \
667              TRANZACT       Fort Lee, NJ     Fort Lee, NJ   
668                  JKGT  San Francisco, CA          Unknown   
669            AccessHope      Irwindale, CA          Unknown   
670  ChaTeck Incorporated  San Francisco, CA  Santa Clara, CA   
671         1-800-Flowers       New York, NY  Carle Place, NY   

                       Size  Founded  Type of ownership  \
667  1001 to 5000 employees     1989  Company - Private   
668                 Unknown       -1            Unknown   
669                 Unknown       -1            Unknown   
670       1 to 50 employees       -1  Company - Private   
671  1001 to 5000 employees     1976   Company - Public   

                    Industry             Sector                 Revenue  \
667  Advertising & Marketing  Business Services                 Unknown   
668                  Unknown            Unknown                 Unknown   
669                  Unknown            Unknown                 Unknown   
670  Advertising & Marketing  Business Services  $1 to $5 million (USD)   
671                Wholesale  Business Services  $1 to $2 billion (USD)   

    Competitors  Min Salary  Max Salary  Average Salary  
667     Unknown      105000      167000        136000.0  
668     Unknown      105000      167000        136000.0  
669     Unknown      105000      167000        136000.0  
670     Unknown      105000      167000        136000.0  
671     Unknown      105000      167000        136000.0

### 'Same State' column

First, let us check whether or not there are any *Location* rows with a value of 'Unknown' (which would be relly sus by the way).

In [70]:
print(job_posting['Location'].isin(['Unknown']).any())

False


There are none, so let us keep going. 

In [91]:
def compare_states(row):
    hq = row['Headquarters']
    location_state = row['Location'][-2:]
    if hq != 'Unknown':
        return True if location_state == hq[-2:] else False
    return False

job_posting['Same State'] = job_posting.apply(compare_states, axis=1)


Code above applies `compare_states` function to each row and stores the result (boolean value) in *Same State* column. 

Let's see how our new column looks like.

In [94]:
job_posting['Same State']

0       True
1       True
2       True
3      False
4       True
       ...  
667     True
668    False
669    False
670     True
671     True
Name: Same State, Length: 672, dtype: bool

In [96]:
# displaying all of the 'Same State' cols values
job_posting['Same State'].unique()

array([ True, False])

### Additional changes

I would also like to see what technologies are most frequently preferred by employers, you know, asking for a friend :)

In [115]:
job_posting['Python'] = job_posting['Job Description'].map(lambda str: 'python' in str.lower())
job_posting['Excel'] = job_posting['Job Description'].map(lambda str: 'excel' in str.lower())
job_posting['SQL'] = job_posting['Job Description'].map(lambda str: 'sql' in str.lower())
job_posting['Spark'] = job_posting['Job Description'].map(lambda str: 'spark' in str.lower())
job_posting['Power BI'] = job_posting['Job Description'].map(lambda str: 'power bi' in str.lower())

Let us have a closer look on those values.

In [120]:
job_posting[['Python', 'Excel', 'SQL', 'Spark', 'Power BI']].apply(lambda col: col.value_counts())

Python  Excel  SQL  Spark  Power BI
False     181    376  316    483       638
True      491    296  356    189        34

I would like to visualize this chart .............

In [121]:
import seaborn as sns

ModuleNotFoundError: No module named 'seaborn'